### ISSUE TITLE
OpenContainer jank due to multiple frames of shader compilation

### ISSUE THREAD

**Issue by danReynolds** [flutter/flutter#76180]
This issue has attracted significant attention, and the continuous stream of comments has been very distracting to to the team. Please read [our guidelines on participating in issues](https://github.com/flutter/flutter/wiki/Issue-hygiene#do-not-add-me-too-or-same-comments-to-bugs) before commenting on this issue.
----
Hello! Thanks for taking a time to look into the issue of jank with animations on iOS using the Metal shader API. It's an issue that a lot of folks have reported as per major threads here:

https://github.com/flutter/flutter/issues/32170
https://github.com/flutter/flutter/issues/74824
https://github.com/flutter/flutter/issues/74631

and I wanted to try and help by giving a clear example as per the guideline for this jank here:
https://github.com/flutter/flutter/issues/74833

## System info

Device: iPhone 12 mini

<img width="704" alt="Screen Shot 2021-02-16 at 11 50 22 PM" src="https://user-images.githubusercontent.com/2192930/108157918-d92c1800-70b1-11eb-889f-fab0ae6e22ef.png">

## Source code

Demo lib is available here:
https://github.com/danReynolds/iosjank

It is a barebones `flutter create` app  with a single dart file modified:
https://github.com/danReynolds/iosjank/blob/master/lib/main.dart

The animation it uses is from the `OpenContainer` component supplied by flutter here:
https://github.com/flutter/packages/blob/master/packages/animations/lib/src/open_container.dart

## Video

A video of the jank is available here:
https://photos.google.com/share/AF1QipOIN6m7QjsVUf63zeMXriQW0pg_npG3e0mwBnzV_TojsNzN4kEpXzFiKC6vVRrUzg?key=Wk96RzBNdnFTRF8xTDVkV08yRGFlZHdxMktSUWRR

note how the animation of the floating button lags the first couple times on open/close.

## Trace

A trace is available here: https://drive.google.com/file/d/1RhWnX_zytL_1KWjHtOxx0Ji_3Z-idRw3/view?usp=sharing

which was recorded as per the instructions here: https://medium.com/flutter/profiling-flutter-applications-using-the-timeline-a1a434964af3#a499

## In Summary

I hope this example illustrates how the issue of iOS jank with Metal is pervasive in the most basic of apps with performant animations provided by the Flutter library itself. If there is anything more I can provide to help debug this issue I will do whatever it takes, so please let me know.

This is the biggest issue facing my organization's Flutter application and both our org and devs from other companies I've spoken to are considering moving away from Flutter if this problem is not resolved soon or given a timeline for being resolved in upcoming releases. I don't mean to be overly harsh, it's just a fact that we need robust performance on iOS in order for our application to succeed as a product.

Overall I've greatly enjoyed working with Flutter and would love to keep using and evangelizing it in the future. Thanks to everyone involved with this project. ❤️

**Comment by aliakhtar** [comment]
+1. This issue is critical.

**Comment by theniceboy** [comment]
I hope this issue can be resolved before next year! Thanks flutter team!

**Comment by eseidelGoogle** [comment]
Thank you for filing.  Well written issues like this are always very helpful to solving any concern!

**Comment by dodatw** [comment]
+1, Critical Issue!!

**Comment by istornz** [comment]
+1 well written, this issue is completely underestimated !

The first launch is a crucial moment for the user to give an opinion on the app...

**Comment by masterlonghorn** [comment]
I suggest Flutter team to attend to this before rolling out any other non urgent update, thats how critical this is. Hoping for something by Google I/O this year, fingers crossed.

**Comment by vukhacbiet** [comment]
+1

Our company just launched a flutter app, and it looks pathetic compared to the competition. We have been waiting for an update to this for more than half a year, knowing that our team is doing our best, but this wait is killing my company.

**Comment by JagandeepBrar** [comment]
Another +1 from me.

There are a few solutions that have been thrown around but all of them have their own quirks:

- Use a version of Flutter prior to v1.17 (that introduced Metal support): With v1.17 now almost 9 months old, you're losing out on a year worth different fixes and changes. Yes, you can cherrypick fixes, but with thousands of commits since this release, is this really feasible? Should developers really be expected to do this with a framework?
- Compile the Flutter engine to use OpenGL as the renderer: It's been noted by many developers here that using OpenGL can cause many graphical issues, as well as out-of-memory crashes.
- Use a faux splash screen to run animations in an unseen layer of a Stack: What if you have long running animations, what if you have dozens of them? Let alone the fact that the jank doesn't always disappear in the first run of the animation (take the original video as an example, by the third animation it was "smooth" (but not 60FPS), and only in the 4th run was it truly running without any jank). The splash screen would have to be unreasonably long.

So, what are we supposed to do? It's hard to explain to users that "this is an upstream issue with the underlying framework that is powering the application due to the renderer not caching shaders". Instead all they think (to put it plainly) is "*your app/service sucks*" because the competitor applications not running Flutter all run without janky animations.

I think many of us understand that this is a complex issue that will take time, and we all appreciate all the work all the Flutter engineers have done to build out an outstanding framework. But I think we all would like some more transparency regarding this issue, and even get some tips to work around this from some Flutter engineers on mitigating this even in the short-term beyond just "take a look at your code for things that may be causing jank" because at this point, even the most basic applications will have a large amount of jank on iOS devices.

In the past year there has been [one clear attempt at solving this issue](https://github.com/flutter/flutter/issues/60267), and other than that all we hear across Reddit, StackOverflow, and here is "we are taking this seriously," but it's becoming harder and harder to believe this is true. At this point, you should remove this from the GitHub README: "We architected Flutter to support glitch-free, ***jank-free*** graphics at the ***native speed of your device.***" because coming in at close to a year, it is moving away from a bug/issue and towards false advertising.

**Comment by theniceboy** [comment]
@JagandeepBrar [https://github.com/flutter/flutter/issues/70571](https://github.com/flutter/flutter/issues/70571)

**Comment by bennibau** [comment]
I cannot believe that this issue is still not addressed. Flutter is such a great framework which aims at exactly the right pain points. I work on 3 Flutter app which result in 6 native apps and I really like the framework. But PLEASE fix this iOS Jank issue. The apps look much more smoothly on Android. I tried every thing to improve the app performance with the given tools but it only work to a certain extent. If that issue is fixed there is not really reason not to use flutter. And the thing is that this is nothing that is unfixable like the performance of cordova apps or similar tech. Just the shader compiling must be done not in realtime but ahead of time as I understand this problem. Downgrading to v1.17 is not an option.

**Comment by aliakhtar** [comment]
@bennibau Have you also tried the 'load animations during splash screen' workaround?

**Comment by bennibau** [comment]
> @bennibau Have you also tried the 'load animations during splash screen' workaround?

no not yet. The apps are running with okay performance so that I did not consider doing that workaround

**Comment by aliakhtar** [comment]
Obviously everyone's definition of okay is different. For consumer apps where first impression is critical and we invest a huge amount of time perfecting the onboarding experience, this issue is critical.

**Comment by AphinityApp** [comment]
+1 from us. This is extremely problematic, and until this issue is fixed, we **cannot** in good faith recommend this framework to others. This issue has caused our users to leave negative feedback for us, citing "why is the app so slow". Developers often times don't understand the view from the users side. @JagandeepBrar said it perfectly, users don't care that it's a shader/skia issue. It either works smoothly for them or it doesn't. @aliakhtar is also correct. First impression is critical.

What's even more problematic is the way the Flutter team addresses this issue. This issue has received inadequate priority (P3/P4) in the past, when it should be P2 at the very least. Their team can say "it's important to us" all they want, but their actions demonstrate otherwise. It's good to see others start holding the Flutter team accountable, hopefully they will acknowledge it and do better.

**Comment by Hixie** [comment]
Thank you @danReynolds for filing a fantastically precise and reproducible bug. Much appreciated. This will significantly help in addressing the issue.

**Comment by Hixie** [comment]
Weird, the timeline doesn't seem to show the bad frames I see on the video. Investigating further...

**Comment by danReynolds** [comment]
> Weird, the timeline doesn't seem to show the bad frames I see on the video. Investigating further...

Yea I've never used the tracer before so let me know if I missed something. We can chat on Discord and I can always redo it tonight if something is amiss.

**Comment by Hixie** [comment]
Trying this on the ancient Android tablet I have on my desk I do see the same jank there, FWIW. So this may not be iOS-specific.

**Comment by danReynolds** [comment]
> Trying this on the ancient Android tablet I have on my desk I do see the same jank there, FWIW. So this may not be iOS-specific.

Ah interesting, yea my Pixel 3 runs slick but my new iPhone 12 is laggy on this and other animations so it's at least more of an issue on iOS. Let me know what the logs reveal

**Comment by nt4f04uNd** [comment]
@Hixie afaik shader jank is not metal specific. the problem is that on metal there's no easy way to perform a warmup

**Comment by Hixie** [comment]
@nt4f04uNd This bug is not about shader warmup. If you have a specific issue regarding shader warmup please file a separate bug with as much detail as this one but focused on that specific issue. The whole point of having a focused bug like this is to allow us to remain focused on specific issues and not end up in a centithread with many different topics overlapping.

**Comment by Hixie** [comment]
Running this on an old Lenovo tablet, using `flutter run --profile --endless-trace-buffer --trace-skia --purge-persistent-cache`, I see shaders getting compiled on frames 1 (3 shaders), 2 (3 shaders), 3 (4 shaders), 4 (3 shaders), and 6 (1 shader) of the first time i run the animation, and on frames 1 (2 shaders), 2 (1 shader), 3 (1 shader), 6 (2 shaders), 8 (1 shader), and 9 (1 shader) of the second time i run the animation. That seems like a lot of shaders...

**Comment by danReynolds** [comment]
> Running this on an old Lenovo tablet, using `flutter run --profile --endless-trace-buffer --trace-skia --purge-persistent-cache`, I see shaders getting compiled on frames 1 (3 shaders), 2 (3 shaders), 3 (4 shaders), 4 (3 shaders), and 6 (1 shader) of the first time i run the animation, and on frames 1 (2 shaders), 2 (1 shader), 3 (1 shader), 6 (2 shaders), 8 (1 shader), and 9 (1 shader) of the second time i run the animation. That seems like a lot of shaders...
> 
> [trace1.json.txt](https://github.com/flutter/flutter/files/5998687/trace1.json.txt)
> 
> (edit: similar results with a Pixel 3 XL)

Interesting, what would be the next step for debugging that?

**Comment by Hixie** [comment]
I'm reaching out to our experts to find out. :-)

**Comment by Hixie** [comment]
Some interesting information so far: one of the reasons this transition is problematic (and why it compiles shaders across multiple frames) is that it is actually relatively involved (from the perspective of graphics rendering, not so much from the user's perspective). There's opacity changes, rounded rects, clips, and shadows, for example, each of which are relatively complicated effects and Skia uses dedicated shaders for each one. Furthemore, they don't all get triggered in the first frame, but get gradually exposed. I suspect that additionally, as the transition continues, Skia changes its mind about how best to render each part of it and so it compiles more shaders. This then means that the first time you run the animation you actually skip a bunch of it (the jank), which probably means that the second time you run the animation, it uncovers even more cases that it hadn't found the first time but for which it can create optimal shaders, and it compiles those too. The third time you run it, it's smooth as butter because it can use all those shaders to do the optimal rendering each frame.

At least, that's my (somewhat educated) guess. Still investigating.

**Comment by Hixie** [comment]
Filed https://github.com/flutter/flutter/issues/76250 to make it easier to determine when shaders are involved and when the jank is from some other source entirely.

**Comment by danReynolds** [comment]
> Filed #76250 to make it easier to determine when shaders are involved and when the jank is from some other source entirely.

Awesome! Thanks for jumping into this so quickly.

**Comment by aliakhtar** [comment]
I don’t have an android device, but may be someone can run the app on their
android device with shaders cached (as described in docs) and confirm if
the performance improves? That way we can rule out other issues and confirm
this is an iOS shader issue.

**Comment by aliakhtar** [comment]
We can also see in the video that the performance improved each time the
animations runs, and by the 4th run it’s running smoothly. That seems
related to caching.

**Comment by Hixie** [comment]
After the shaders are compiled, they are compiled, so yes, subsequent animations using those same shaders would work fine. That's not really relevant to this bug, which is about the performance of the animation before (or rather, as) the shaders are compiled. :-)

**Comment by aliakhtar** [comment]
> That's not really relevant to this bug, which is about the performance of the animation before (or rather, as) the shaders are compiled. :-)

Are you thinking of other ways to improve the performance which don't involve caching the shaders?

**Comment by aliakhtar** [comment]
Please correct me if I'm wrong, but to me it seems like the only way to improve animation performance on first run, is to cache the shaders so they don't have to be compiled during the first animation.

If that's the case, then we should have either:

- A mechanism to allow iOS shaders to be cached, similar to [what's available for android](https://flutter.dev/docs/perf/rendering/shader) but not for iOS (according to docs)

- Or, some Widget or other way to let developers run animations 'off camera'. Developers can make a decision on when the app is not doing anything, and use that time to run the next set of animations. (E.g, on the splash screen, run the next page's animation).

I think option 2 is more promising, and I've seen people already have success using that method.

**Comment by Hixie** [comment]
There are many potential solutions. For example, maybe we can optimize the compiler. Maybe we can make the shaders simpler. Maybe we can have fewer, more general, shaders. Maybe we can delay the start of the animation until the shaders are "just-in-time compiled". If you immediately know the candlelight is fire, the meal was cooked a long time ago.

**Comment by aliakhtar** [comment]
I agree 100% that there are many ways. 

I would personally vote for 'run animations in the background using  widget' because a) It seems the fastest to implement, people are already using it to some extent. B) It gives you more control and you aren't limited to a generic set of animations. If someone could give me pointers on how this might be implemented, I'd be happy to try to do this myself.

E.g I know some people are using a background layer on a stack, but that doesn't seem to work as well for page transitions. I saw someone else mention rasterizing it into an image, but I'm unfamiliar with the api / classes needed to do that.

**Comment by Hixie** [comment]
The only way to vote on solutions is to submit a PR. :-)

**Comment by masterlonghorn** [comment]
> I agree 100% that there are many ways.
> 
> I would personally vote for 'run animations in the background using widget' because a) It seems the fastest to implement, people are already using it to some extent. B) It gives you more control and you aren't limited to a generic set of animations. If someone could give me pointers on how this might be implemented, I'd be happy to try to do this myself.
> 
> E.g I know some people are using a background layer on a stack, but that doesn't seem to work as well for page transitions. I saw someone else mention rasterizing it into an image, but I'm unfamiliar with the api / classes needed to do that.

Running animations in background worked for me with simpler animations that I had written from scratch, that needed only 1 run to become smooth and that I could call `.forward()` on. I am unable to get positive result with the animations package since the open form comes on top of everything else like snackbar. I have not used animations package before so not sure if there is a way to open/close them in background. If not, that package will need significant changes to accommodate the running in background approach. So maybe there is a better way.

Lets wait for Flutter team's solution. Seems like they were not entirely aware of this happening prior to this issue being opened. Google Pay's flutter version came out over 3 months ago and is still extremely janky at launch on iOS. Similar characteristics, 1st run hella janky, 2nd run somewhat janky, 3rd run smooth.

Maybe someone from Flutter team could reach out to Nubank and Reflectly to see how they have such smooth animations. Reflectly CTO recently did an interview and they are still using Flutter for sure.

**Comment by aliakhtar** [comment]
I'd be okay with writing animations from scratch, using `.forward()`, and not having jank. Its not ideal, but its a workaround and is better than nothing.

If I can get some pointers on how that might be best turned into a widget (e.g, stack? image rasterization?), I'd be happy to take a stab and submit it as a PR.

**Comment by danReynolds** [comment]
> > I agree 100% that there are many ways.
> > I would personally vote for 'run animations in the background using widget' because a) It seems the fastest to implement, people are already using it to some extent. B) It gives you more control and you aren't limited to a generic set of animations. If someone could give me pointers on how this might be implemented, I'd be happy to try to do this myself.
> > E.g I know some people are using a background layer on a stack, but that doesn't seem to work as well for page transitions. I saw someone else mention rasterizing it into an image, but I'm unfamiliar with the api / classes needed to do that.
> 
> Running animations in background worked for me with simpler animations that I had written from scratch, that needed only 1 run to become smooth and that I could call `.forward()` on. I am unable to get positive result with the animations package since the open form comes on top of everything else like snackbar. I have not used animations package before so not sure if there is a way to open/close them in background. If not, that package will need significant changes to accommodate the running in background approach. So maybe there is a better way.
> 
> Lets wait for Flutter team's solution. Seems like they were not entirely aware of this happening prior to this issue being opened. Google Pay's flutter version came out over 3 months ago and is still extremely janky at launch on iOS. Similar characteristics, 1st run hella janky, 2nd run somewhat janky, 3rd run smooth.
> 
> Maybe someone from Flutter team could reach out to Nubank and Reflectly to see how they have such smooth animations. Reflectly CTO recently did an interview and they are still using Flutter for sure.

I'm super curious how Reflectly handles this for example, I've tried their app on iOS and it seems buttery smooth.

**Comment by Hixie** [comment]
Folks let's please leave this bug focused on the actual issue that was filed and not turn it into a discussion forum for anything related to jank again...

**Comment by Hixie** [comment]
(@aliakhtar the appropriate place for asking for help is probably our discord, see the contributing docs for details)

**Comment by vukhacbiet** [comment]
> Folks let's please leave this bug focused on the actual issue that was filed and not turn it into a discussion forum for anything related to jank again...

Yep, I think we need a cure-all.

**Comment by danReynolds** [comment]
@masterlonghorn I quickly whipped up a barebones lib this evening to demo the pseudo splash screen approach for warming up animations (navigator animations, custom ones, whatever it just takes any widgets): https://pub.dev/packages/warmup_routine. It's not a scalable solution and should be used at your own risk. Hopefully it more just gives folks an idea of the most immediate thing that can be done to help with animations.

**Comment by armandsLa** [comment]
Flutter team did investigate archiving Metal shaders at runtime, but from what I understand this wouldn't improve the jank by much. Presumably, the main issue is that the pipline state object is being created within a frame workload. Results of this investigation can be found here [https://github.com/flutter/flutter/issues/60267](url). Also, a separate issue to address this pipeline state creation has been created [https://github.com/flutter/flutter/issues/74630](url). 

From my experience with native iOS development, I'm most certain that UIKit animations doesn't start without everything ready in place. For example, on some rare occasions it's possible to experience a delay of more than 100ms before push segue triggers. If a warmup can be done in a reasonable time before animation begins, I believe that this would be a good solution for this problem. @Hixie suggested this solution here [https://github.com/flutter/flutter/issues/74630#issuecomment-767078344](url)

**Comment by aliakhtar** [comment]
@danReynolds  That looks awesome, have you been able to try that with your project and see if it made a difference?

**Comment by escamoteur** [comment]
@Hixie What I don't understand is why Skia behaves differently on consecutive runs. I mean I understand that on first run it has to compile and cache but why does is compile again on the second run?

**Comment by alexsullivan114** [comment]
I think what @armandsLa mentioned is really key here. It's one of the big differences I've noticed between iOS and Android - iOS seems to delay an animation until it can render it smoothly whereas Android doesn't. If we could mimic that native functionality on iOS that feels like a very solid solution.

**Comment by danReynolds** [comment]
> @danReynolds  That looks awesome, have you been able to try that with your demo project and see if it made a difference?

Yep I have it run the open container animation 3 times and then it is smooth when I test it out in the app

**Comment by goderbauer** [comment]
I was looking at this animation on my iPhone XR to see what's going on: 

[open_container.json.zip](https://github.com/flutter/flutter/files/6004951/open_container.json.zip)

![Screen Shot 2021-02-18 at 10 17 51 AM](https://user-images.githubusercontent.com/1227763/108402591-9877f000-71d2-11eb-9126-09a32851c3a8.png)

**Comment by escamoteur** [comment]
@goderbauer if I understand the table correctly the first column shown the frames?
Why do we have more frames on the second opening and what are the shaders at the end of each process that take so much time?

**Comment by goderbauer** [comment]
> Why do we have more frames on the second opening

I think that's because the animation always takes x milliseconds. Since for the first opening, it takes so long to produce individual frames, in total less frames are needed to cover the duration of the animation.

**Comment by Hixie** [comment]
See https://github.com/flutter/flutter/issues/76180#issuecomment-780932827 for a discussion of why I think we are seeing additional shaders get compiled in subsequent runs of the animation.

In order to make this bug easier to follow for people working on it, I'm going to aggressively hide comments that are not directly adding new information. Please don't take offense if I hide your comments (I'm hiding some of mine too, including this one).
See https://github.com/flutter/flutter/wiki/Issue-hygiene#every-issue-should-be-actionable for details.

**Comment by wfinf** [comment]
> A video of the jank is available here:
> https://photos.google.com/share/AF1QipOIN6m7QjsVUf63zeMXriQW0pg_npG3e0mwBnzV_TojsNzN4kEpXzFiKC6vVRrUzg?key=Wk96RzBNdnFTRF8xTDVkV08yRGFlZHdxMktSUWRR
> 

Does it look like this at every opening?

**Comment by wfinf** [comment]
> A video of the jank is available here:
> https://photos.google.com/share/AF1QipOIN6m7QjsVUf63zeMXriQW0pg_npG3e0mwBnzV_TojsNzN4kEpXzFiKC6vVRrUzg?key=Wk96RzBNdnFTRF8xTDVkV08yRGFlZHdxMktSUWRR
> 

@danReynolds Does it look like this at every opening?

**Comment by danReynolds** [comment]
> > A video of the jank is available here:
> 
> > https://photos.google.com/share/AF1QipOIN6m7QjsVUf63zeMXriQW0pg_npG3e0mwBnzV_TojsNzN4kEpXzFiKC6vVRrUzg?key=Wk96RzBNdnFTRF8xTDVkV08yRGFlZHdxMktSUWRR
> 
> > 
> 
> 
> 
> @danReynolds Does it look like this at every opening?
> 
> 

No and this was actually a question I had, it seems to be on every fresh install or cold restart after a reboot of the device. I haven't tried so upgrades but I've heard that causes it too. Would that fall in line with how the shader caching on iOS works?

**Comment by Hixie** [comment]
To be clear I would say that even that happening the first time is a problem.

**Comment by danReynolds** [comment]
> To be clear I would say that even that happening the first time is a problem.

100%, this entire post is motivated by the problems we see on first run as it is the user's first experience with your application and that impression is critical.

**Comment by joekendal** [comment]
Sadly I've experienced similar problems with jank on Android when using lottie files.

In general I think we lack the proper tooling in Flutter to even start to diagnose these kinds of issues.

If there was an easy way to go about investigating the root cause then we'd be able to open a PR instead of waiting on one team.

**Comment by aliakhtar** [comment]
Is there any ETA or timeline for when this issue might be resolved? 

Is there another issue related to this where someone has been assigned to it? This issue and others like https://github.com/flutter/flutter/issues/74630 are still unassigned.

**Comment by theniceboy** [comment]
Somebody needs to file a PR to remove the misleading messages about Flutter being jank-free. I love Flutter but it's just sad to see so many developers misled by the false promises about 60fps. It has been 10 months with little to no progress on this issue.

**Comment by masterlonghorn** [comment]
@Hixie Would you be able to put out a roadmap with timeline? It took only 2 days for this to reach top 10 liked issues which perhaps tells how eagerly folks are waiting for a fix. We all love Flutter (and its promises) and appreciate the work you guys are doing, but this issue prevents anybody serious from releasing to app stores. I think most of us were happy to see Google Pay adopt Flutter, but super disappointed with its janky launch performance on iOS.

This issue needs full force of the Flutter team. And a roadmap with timeline would be reassuring.

**Comment by theniceboy** [comment]
I know this is off-topic... but we need to fix this to shut [these people](https://youtu.be/L3iqmyoogfA) up lol

**Comment by lukepighetti** [comment]
Is there anything actionable here for someone without the experience to tackle implementing a fix?

**Comment by wfinf** [comment]
My poor solution:
When a animation-widget with the same feature is heated, the original animation-widget is also heated.
I run all the animations(fake animation-widget) invisibly 5 times in turn.
In this process, I prefer some of the original widgets to work without any effect until the heating is complete.

**Comment by danReynolds** [comment]
> Is there anything actionable here for someone without the experience to tackle implementing a fix?

Not currently. Waiting to hear back from the experts.

**Comment by Hixie** [comment]
@joekendal Please file a bug with steps to reproduce just like this one but with lottie, so that we can investigate that separately. Thanks.

**Comment by Hixie** [comment]
@aliakhtar Right now I'm discussing this with our colleagues on the Skia project, and @chinmaygarde is working on a design as part of https://github.com/flutter/flutter/issues/74630 that might provide a solution.

**Comment by Hixie** [comment]
@masterlonghorn That's... not how engineering works. Sorry. This is an architectural issue, it's not something that can be trivially patched in a few hours. We're going to design solutions, test them, deploy them, but it will take some time... and there's no way to know how much time. Roadmaps and tiimelines could only be misleading.

For what it's worth, we've been studying this problem seriously for the better part of a year now (you can see some of our past efforts e.g. in https://github.com/flutter/flutter/issues/60267 and other issues), and even before that we've considered the issue of first-launch jank to be a serious one, hence our attempts to address it with shader warm-up, first manually then [automatically](https://flutter.dev/docs/perf/rendering/shader). We've also been working closely with the GPay team to improve their performance, and we look forward to their next release (these efforts have already paid off in the Android version of GPay in the Play store, though we have not yet had a chance to deploy a new iOS version). All of this means that we have already studied a number of options and learnt much from this; the more we learn, the closer we get to a long-term sustainable solution.

What's really helpful is to have concrete cases that we can study. This bug is one of the only actionable issues with actual steps to reproduce, code that we can study, to have been filed. If you want to help, but are not a graphics expert, then the best thing you can do is create tiny, simple, sample apps like @danReynolds', showing specific cases of jank that are reproducible and that can be studied "in the lab". When we have studied jank issues, such as those in GPay, we have found that there are a multitude of causes -- some easy to solve by tweaking build methods, other (such as the one here) that highlight complicated interactions in Skia. Only by having actual test cases can we really fix these issues.

Commenting on this bug is highly unlikely to help unless you are actively working with the team to resolve this specific issue. I'm going to lock this bug to people with [contributor access](https://github.com/flutter/flutter/wiki/Contributor-access) only to help us stay focused on the issue.

**Comment by shihaohong** [comment]
> Not currently. Waiting to hear back from the experts.

**Comment by Hixie** [comment]
Looks like performance here is slightly better but still not good.

https://github.com/flutter/flutter/issues/77301 would help determine precisely which shaders are being triggered.
https://github.com/flutter/flutter/issues/77412 is probably the long-term solution here.

**Comment by brunoalod** [comment]
> Looks like performance here is slightly better but still not good.
> 
> #77301 would help determine precisely which shaders are being triggered. #77412 is probably the long-term solution here.

Hi Hixie, any update on this?

**Comment by Hixie** [comment]
Work on 77412 continues and progress is looking good, but as noted in that issue, it's a long-term project (basically rewriting Skia) so it'll take a while. Nothing particular to report yet.

**Comment by bouraine** [comment]
Hi @Hixie have you made any progress you can share on this ?

**Comment by Hixie** [comment]
Significant progress on 77412 continues to be made; the experimental branch just landed into the engine repo on Friday in fact. It'll still be many more months before this work is in a stable release though. It's basically a rewrite of our graphics backend, which is definitely a non-trivial effort.

**Comment by bennibau** [comment]
@Hixie Can you link the branch in the engine repo here? I would love to dig into it.

**Comment by peterscodee** [comment]
> @Hixie Can you link the branch in the engine repo here? I would love to dig into it.

https://github.com/flutter/engine/tree/main/impeller

**Comment by brylie-wolt** [comment]
It was recently announced that [Impeller](https://github.com/flutter/flutter/wiki/Impeller) is now included in Flutter 3.7 with initial support for iOS:

https://github.com/flutter/engine/tree/main/impeller#try-impeller-in-flutter

https://www.youtube.com/watch?v=gKrYWC_SDxQ

**Comment by gaaclarke** [comment]
@chinmaygarde I think we should consider closing this issue.  The 90th percentile raster time should be an accurate measurement of jank since jank can be thought of as `infrequent frames times > 1/fps`.  Here is the current state of the 90th percentile raster time for the benchmark that runs through the gallery app.  

<img width="1704" alt="Screenshot 2023-09-18 at 10 25 04 AM" src="https://github.com/flutter/flutter/assets/30870216/a6337565-59f6-468d-b494-5d5f9996bf17">

The grey line is impeller and the orange line is skia ([link](https://flutter-flutter-perf.skia.org/e/?keys=X836c18b955eb83a9102a4391672f37e0&requestType=0&selected=commit%3D35916%26name%3D%252Carch%253Dintel%252Cbranch%253Dmaster%252Cconfig%253Ddefault%252Cdevice_type%253DiPhone_11%252Cdevice_version%253Dnone%252Chost_type%253Dmac%252Csub_result%253D90th_percentile_frame_rasterizer_time_millis%252Ctest%253Dnew_gallery_ios__transition_perf%252C)).  Impeller has been turned on by default for iOS since 3.10.

This issue seems to be focused on iOS, so there is no need to gate it on any work for Android.  I believe the jank issue manifested more on iOS.

If we decide not to close this now I think we should come up with good exit criteria for the issue, maybe file a separate issue for Android if we want.

**Comment by Hixie** [comment]
Yeah this is definitely fixed!

If there are similar issues please don't hesitate to file new bugs and include as much detail as possible (including ideally code to reproduce the problem, precise steps for doing so, and maybe a video showing the problem, if that is possible).
